### DISCUSSION TITLE
Save variable of type array filtering data from data bucket

### DISCUSSION THREAD

**Discussion by LukMas** [mockoon/mockoon#1416] [Help] [answered]
Hi!

I've a data bucket with some data organized into an array. I would like to filter it out and use the result for an array property of my json response. I use [JSON:API](https://jsonapi.org) as schema for my data response. 

I've managed to have what I want with

```
...
  "data": [
    {{#each (dataRaw 'l4yl') }}   
      {{setVar 'currentTemplate' this }}
      {{#each @currentTemplate.data }}
        {{#if (eq this.id (urlParam 'projectId') ) }}
          {
            "id": "{{@currentTemplate.id}}",
            "type": "{{@currentTemplate.type}}"
          },
        {{/if}}
      {{/each}} 
    {{/each}}
  ]
...
```
but the json generated from this is not totally compliant to the specification due to the ',' after the last item. 

```
...
  "data": [
    {
      "id": "100_1",
      "type": "template"
    },
    {
      "id": "100_2",
      "type": "template"
    },
    {
      "id": "100_3",
      "type": "template"
    }, <--- THIS SHOULD NOT BE PRESENT   
  ]
...
```
Then I've tried using a variable. I've had to use a global one because the local one is redefined by a new `setVar` within the loop and not just updated. 

```
...
  "data": [
    {{#each (dataRaw 'l4yl') }}   
      {{setVar 'currentTemplate' this }}
      {{#each @currentTemplate.data }}
        {{#if (eq this.id (urlParam 'projectId') ) }}
          {{#if (getGlobalVar 'afterFirstItem')}},{{/if}}
          {
            "id": "{{@currentTemplate.id}}",
            "type": "{{@currentTemplate.type}}"
          }
          {{setGlobalVar 'afterFirstItem' true }} <--- THIS WITH setVar IS NOT WORKING, IT CREATES A NEW INSTANCE
        {{/if}}
      {{/each}} 
    {{/each}}
  ]
...
```
With this solution the json is compliant. 

Anyway, I would like to know if: 
- would it be possible to update the value of the local variable within the loop? And yes, how? 
- would it be possible to create an array of objects before to use later for the generation of the json? 

Something like this, for example: 

```
{{setVar 'templates' array 
  {{#each (dataRaw 'l4yl') }}
    {{setVar 'currentTemplate' this }}
    {{#each @currentTemplate.relationships.project.data }}
      {{#if (eq this.id (urlParam 'projectId') ) }}
      (object id=@currentTemplate.id type=@currentTemplate.type)
      {{/if}}
    {{/each}} 
  {{/each}}
}}
```
This doesn't work, I know because I've tried. But something similar is working: 

```
{{setVar 'templates' (array (object id='1' type='type1') (object id='2' type='type2')) }}

/// somewhere later
{{{stringify @templates}}}
```

Thank you!

**Comment by 255kb** [comment]
Hi, 

If it's a matter of getting rid of the trailing comma which is indeed invalid JSON, you have to use the [`unless` block helper](https://handlebarsjs.com/guide/builtin-helpers.html#unless). It's part of Handlebars and often used combined with `each`/`@last` (last is a variable available in the each):

```handlebars
{{#each ...}}
  ...
  {{#unless @last}},{{/last}}
{{/each}}
```

The [`repeat` helper](https://mockoon.com/docs/latest/templating/mockoon-helpers/#repeat) we added doesn't have this issue.

Regarding your question with `setVar`, if I remember correctly, Handlebars block helpers (like `each`) create a new context during each iteration, so the variable will be available during this iteration and cannot be overriden.

**Comment by LukMas** [comment]
Hi, 

Thank you for the answer. 
I've tried already with #unless but it's not working. The problem is that there are two loops, and the last value to be printed is not always the last of the loop. 
It works with the global variable, but I think it would be simpler to use with a construct that allows to build an array. Or, to update the content of a variable of type array.

**Reply by 255kb** [reply]
I'm not sure I follow what is the issue without having a full reproducible setup (mockoon setup, request, etc).

But when you use nested `each` you can access the upper level context by using Handlebars: `{{#unless ../last}},{{/unless}}`

**Reply by LukMas** [reply]
I've attached part of the API that I'm mocking. 
[JSON_API.json](https://github.com/user-attachments/files/15909488/JSON_API.json)

The url `localhost:3000/projects` returns the list of the _projects_. For every _project_ you have a list of _templates_ under the `relationship` attribute. I've add an extra attribute, `info`, that is handled with `unless` and `last`. 
As you see, the behaviour is not correct because the last item of the inner array (stored in the `data` attribute) is not (always - in some situations it could be) the last of the outermost loop. 

As you see, it's possible to manage it, but it requires some 'hops' because there is no way to store the list upfront and the variable must be global. A construct that allows to build a list of objects from the content of a data bucket, or a way to add items to a variable of type array, would make the things simpler to understand.

**Reply by 255kb** [reply, marked as answer]
I think I kind of get what you are trying to do but your template is really hard to follow. 
Indeed I see the `"info"` property everywhere, even on the last one, but this is how handlebars works. Filtering with an `if` won't change anything to the `@last` variable. If you are going through 10 items, displaying only the first one, it is not the `@last` one. You will have to handle that differently. 
There is a [`filter` helper](https://mockoon.com/docs/latest/templating/mockoon-helpers/#filter) that could help.

Regarding this: 

> A construct that allows to build a list of objects from the content of a data bucket, or a way to add items to a variable of type array, would make the things simpler to understand.

I'm not sure what you mean. `setVar` and `setGlobalVar` support storing arrays. And for building a list of objects from the content of a bucket, I think the `filter` could work.

If your ultimate goal is to simulate some JSON:API construct with Mockoon, I think it may not be the right tool and coding something custom would be faster that bending Handlebars to your will 😄

**Reply by LukMas** [reply]
Oh... I think that the `filter` helper would fit to my needs. I've totally missed it. :facepalm: 
Given that at any time I know upfront what are the filter conditions, I could get the array of results and use it later when I build the response. 

> `setVar` and `setGlobalVar` support storing arrays.
About this, I know I to build a variable that contains an array of objects, but I've not find a way (if it's possible) to modify it adding a new item. But, as stated above, it looks that the filter is doing what I need. 

> If your ultimate goal is to simulate some JSON:API construct with Mockoon, I think it may not be the right tool and coding something custom would be faster that bending Handlebars to your will

I don't know. As programmers we all try to finds ways to automate our work and, finally, work less to complete our work. Mockoon is really a nice software that has saved us a lot of time to develop our APIs. If I've to code something by myself I could spend a lot of time without matching what you've done up to now. 

Hence, for now it works as we need and we could 'bend' the response to our needs. When this is not working anymore, we will see. 

Thank you for  the answer. `Filter` was the solution.
